File: week1/cash/cash.c
/**
 * @file cash.c
 * @author Ferdi Kanat
 * @date 2025/02/07
 * @brief Calculates the minimum number of coins needed for change
 *
 * This program prompts the user for a cash amount and calculates
 * the minimum number of coins (quarters, dimes, nickels, pennies)
 * needed to make that change.
 *
 * @example
 * Input: Change owed: 0.41
 * Output: 4
 */
#include <stdio.h>
#include <cs50.h>
#include <math.h>

// Constants
#define QUARTER 25
#define DIME 10
#define NICKEL 5
#define PENNY 1
#define NUM_DENOMINATIONS 4

// Function prototypes
static float get_valid_cash(void);
static void calculate_change(float cash);

int main(void)
{
    printf("Welcome to Cash Change Calculator!\n\n");
    float cash = get_valid_cash();
    calculate_change(cash);
    return 0;
}

/**
 * Gets a valid cash amount from the user.
 * @return Float value greater than 0
 * @note The function will keep prompting the user until a valid cash amount is entered.
 * @note The function uses get_float() from the CS50 library.
 */
static float get_valid_cash(void)
{   
    float cash;
    do {
        cash = get_float("Change owed: ");
    } while (cash <= 0);
    return cash;
}

/**
 * Calculates the change owed to the customer.
 * @param cash The cash amount entered by the user
 */
static void calculate_change(float cash)
{
    int cents = round(cash * 100);
    int coin_count = 0;
    int denominations[] = {QUARTER, DIME, NICKEL, PENNY};
    for (int i = 0; i < NUM_DENOMINATIONS; i++)
    {
        coin_count += cents / denominations[i];
        cents %= denominations[i];
    }
    printf("%d\n", coin_count); // Added newline for clean output
}

File: week6/cash/cash.py
"""
Cash - A program that calculates the minimum number of coins required to give a user change.
CS50x Kodluyoruz Week 6 Problem
"""

from cs50 import get_float
from typing import List

# Coin denominations in cents (quarters, dimes, nickels, pennies)
DENOMINATIONS = [25, 10, 5, 1]


def main() -> None:
    """
    Main function that prompts for change amount and displays minimum coins needed.
    """
    # Get the amount of change owed from the user
    change_amount = get_change_amount()

    # Convert the amount to cents
    cents = round(change_amount * 100)

    # Calculate and print the minimum number of coins
    coin_count = calculate_min_coins(cents, DENOMINATIONS)
    print(coin_count)


def get_change_amount() -> float:
    """
    Prompts the user for a positive amount of change.

    Returns:
        float: The change amount in dollars
    """
    while True:
        try:
            change_owed = get_float("Change owed: ")
            if change_owed > 0:
                return change_owed
            print("Please enter a positive amount.")
        except ValueError:
            print("Please enter a valid numeric amount.")


def calculate_min_coins(amount: int, coin_denominations: List[int]) -> int:
    """
    Calculates the minimum number of coins needed for the given amount.

    Args:
        amount: The amount in cents
        coin_denominations: Available coin denominations in cents

    Returns:
        int: Minimum number of coins needed
    """
    coins = 0

    for coin in coin_denominations:
        coins += amount // coin
        amount %= coin

    return coins


if __name__ == "__main__":
    main()